applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,264 @@
+ * The intervals are encoded by OR operation with a magic number of 0x0d00.
+ */
+#define REG_CONN_INTERVAL_LLPM_MASK	0x0d00
+
+enum state {
+STATE_INITIALIZING,
+STATE_READY,
+STATE_ERROR,
+STATE_COUNT
+}
+
+static enum state module_state;
+
+enum dvfs_state {
+	DVFS_STATE_INITIALIZING,
+	DVFS_STATE_LLPM_CONNECTED,
+	DVFS_STATE_USB_CONNECTED,
+	DVFS_STATE_COUNT
+};
+
+static const uint8_t dvfs_high_freq_bitmask = BIT(DVFS_STATE_USB_CONNECTED) |
+					      BIT(DVFS_STATE_INITIALIZING);
+static const uint8_t dvfs_medlow_freq_bitmask = BIT(DVFS_STATE_LLPM_CONNECTED);
+
+/* Binary mask tracking which states are requested.
+ * We start with DVFS_STATE_INITIALIZING on as it is active on start.
+ */
+static uint8_t dfvs_requests_state = BIT(DVFS_STATE_INITIALIZING);
+
+static enum dvfs_frequency_setting current_freq = DVFS_FREQ_HIGH;
+
+static struct dvfs_retry_work_data {
+	struct k_work_delayable dvfs_retry_work;
+	uint8_t dfvs_retries_cnt;
+} dvfs_retry_data;
+
+static const char *get_dvfs_frequency_setting_name(enum dvfs_frequency_setting setting)
+{
+	switch (setting) {
+	case DVFS_FREQ_HIGH: return "DVFS_FREQ_HIGH";
+	case DVFS_FREQ_MEDLOW: return "DVFS_FREQ_MEDLOW";
+	case DVFS_FREQ_LOW: return "DVFS_FREQ_LOW";
+	default: return "Unknown";
+	}
+}
+
+static const char *get_dvfs_state_name(enum dvfs_state state)
+{
+	switch (state) {
+	case DVFS_STATE_INITIALIZING: return "DVFS_STATE_INITIALIZING";
+	case DVFS_STATE_LLPM_CONNECTED: return "DVFS_STATE_LLPM_CONNECTED";
+	case DVFS_STATE_USB_CONNECTED: return "DVFS_STATE_USB_CONNECTED";
+	default: return "Unknown";
+	}
+}
+
+static void reschedule_dfvs_retry_work(enum dvfs_frequency_setting target_freq, size_t dvfs_ret_val)
+{
+	if (dvfs_retry_data.dfvs_retries_cnt > DVFS_NUMBER_OF_RETRIES) {
+		LOG_ERR("DVFS retry count exceeded.");
+		module_set_state(MODULE_STATE_ERROR);
+		module_state = STATE_ERROR;
+		return;
+	}
+	if (k_work_delayable_is_pending(&dvfs_retry_data.dvfs_retry_work)) {
+		return;
+	}
+	dvfs_retry_data.dfvs_retries_cnt++;
+	k_timeout_t timeout;
+
+	if (dvfs_ret_val == -EBUSY) {
+		LOG_DBG("DVFS frequency change in progress");
+		timeout = DVFS_RETRY_TIMEOUT;
+	} else if (dvfs_ret_val == -EAGAIN) {
+		LOG_DBG("DVFS not initialized, trying again.");
+		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
+	} else {
+		LOG_ERR("DVFS freq change returned with error: %d", ret);
+		module_set_state(MODULE_STATE_ERROR);
+		module_state = STATE_ERROR;
+		return;
+	}
+	(void) k_work_reschedule(&dvfs_retry_data.dvfs_retry_work, timeout);
+}
+
+static void cancel_dvfs_retry_work(void)
+{
+	(void) k_work_cancel_delayable(&dvfs_retry_data.dvfs_retry_work);
+	dvfs_retry_data.dfvs_retries_cnt = 0;
+}
+
+static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+{
+	if (module_state == STATE_ERROR) {
+		return;
+	}
+
+	size_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+
+	if (ret) {
+		reschedule_dfvs_retry_work(target_freq, ret);
```

Comment:
Can we call it `handle_dvfs_error` ?

Reply:
I prefer current name

Reply:
@mkapala-nordic what's your opinion on this? Handling error codes that do not lead to work rescheduling in `reschedule_dfvs_retry_work` is a bit fishy

Reply:
Marek's suggestions looks good to me, it explains better what is our intention here.
Also we reschedule on only 2 error codes, for the other we do not do that so it is misleading.

> it suggest we encounter an error and it is sth critical

IMO an error is an error, no matter the severity.

Reply:
> @mkapala-nordic what's your opinion on this? Handling error codes that do not lead to work rescheduling in `reschedule_dfvs_retry_work` is a bit fishy

I can always change back to first implementation there you won't have handling error codes that do not lead to work rescheduling